packages/router/src/router.tsx

```diff
@@ -1,41 +1,221 @@
+/**
+ * External dependencies
+ */
+import RouteRecognizer from 'route-recognizer';
+import { createBrowserHistory } from 'history';
+
 /**
  * WordPress dependencies
  */
 import {
 	createContext,
 	useContext,
 	useSyncExternalStore,
+	useMemo,
 } from '@wordpress/element';
+import {
+	addQueryArgs,
+	getQueryArgs,
+	getPath,
+	buildQueryString,
+} from '@wordpress/url';
 
 /**
  * Internal dependencies
  */
-import history from './history';
-import type { EnhancedHistory } from './history';
+import type { ReactNode } from 'react';
+
+const history = createBrowserHistory();
+interface Route {
+	name: string;
+	path: string;
+	areas: Record< string, ReactNode >;
+	widths: Record< string, number >;
+}
 
-const RoutesContext = createContext< Location | null >( null );
-const HistoryContext = createContext< EnhancedHistory >( history );
+type LocationWithQuery = Location & {
+	query?: Record< string, any >;
+};
+
+interface Match {
+	name: string;
+	path: string;
+	areas: Record< string, ReactNode >;
+	widths: Record< string, number >;
+	query?: Record< string, any >;
+	params?: Record< string, any >;
+}
+
+export type BeforeNavigate = ( arg: {
+	path: string;
+	query: Record< string, any >;
+} ) => {
+	path: string;
+	query: Record< string, any >;
+};
+
+interface Config {
+	pathArg: string;
+	beforeNavigate?: BeforeNavigate;
+}
+
+export interface NavigationOptions {
+	transition?: string;
+	state?: Record< string, any >;
+}
+
+const RoutesContext = createContext< Match | null >( null );
+export const ConfigContext = createContext< Config >( { pathArg: 'p' } );
```

Comment:
The `pathArg` field could be generalized into something called `pathSource`. And support getting path both from query string and the actual path. You have something similar implemented already in the `add_rewrite_rule` commit. Options:
1. `pathSource={ query( 'p' ) }` (or `searchParams( 'p' )`) gets the path from query string
2. `pathSource={ pathname( { basePath: ... } ) }` gets it from `location.pathname`, relative to optional `basePath`. If `basePath` is `/nested-site/wp-admin/design/styles`, then the resulting `path` should be just `/styles`.

A config like this easily supports all possible scenarios.

Reply:
Yes, and in both directions the algorithm would be different. The getter reads from the `p` query arg that's been already rewritten, but the setter needs to construct a pretty URL.